### PR TITLE
segmental conll 2000 dataset reader

### PR DESCRIPTION
**Pull request by swabhs**
This pull request includes
- a new dataset reader for conll 2000, which creates fields necessary for a segmental language model.
- test for the dataset reader
- test fixtures
- minor updates to the evaluate command in allennlp.

### CODE REVIEW

File: allennlp/commands/evaluate.py [hunk 1]
```diff
@@ -196,7 +197,8 @@ def evaluate_from_args(args: argparse.Namespace) -> Dict[str, Any]:
         logger.info("%s: %s", key, metric)
 
     output_file = args.output_file
-    if output_file:
-        with open(output_file, "w") as file:
-            json.dump(metrics, file, indent=4)
+    if not output_file:
+        output_file = os.path.join(args.archive_file.split("model.tar.gz")[0], 'metrics.json')
```
Comment (brendan-ai2) [inline allennlp/commands/evaluate.py]: This sort of feature I would implement on the command line (or in a script). Something like:
`export MY_DIR=my/path/to/model && allennlp evaluate --archive_file=$MY_DIR/model.tar.gz --output-file=$MY_DIR/metrics.json --other_args=foo`

Comment (brendan-ai2) [reply allennlp/commands/evaluate.py]: That keeps the interface to evaluate standard meaning fewer merge conflicts. Also, you can use CTRL-R on the command line to quickly recall such commands. So it wouldn't mean more typing either! :)

File: allennlp/data/dataset_readers/segmental_conll2000.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        file_path = cached_path(file_path)
+
+        with open(file_path, 'r') as data_file:
+            logger.info('Reading instances from lines in file at: %s', file_path)
+
+            # Group into alternative divider / sentence chunks.
+            for is_divider, lines in itertools.groupby(data_file, _is_divider):
+                # Ignore the divider chunks, so that `lines` corresponds to the words
+                # of a single sentence.
+                if not is_divider:
```
Comment (brendan-ai2) [inline allennlp/data/dataset_readers/segmental_conll2000.py]: Tip: In imperative languages a lot of people prefer saying

```
if is_divider:
    continue

fields = [line.strip...
```

This keeps the code a bit flatter which can improve readability.

File: allennlp/data/dataset_readers/segmental_conll2000.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                    fields = [line.strip().split() for line in lines]
+                    # unzipping trick returns tuples, but our Fields need lists
+                    tokens, _, chunk_tags = [list(field) for field in zip(*fields)]
+
+                    # HACK: Ignore long training sentences for memory efficiency.
+                    if "train" in file_path and len(tokens) > self._max_train_sentence_len:
```
Comment (brendan-ai2) [inline allennlp/data/dataset_readers/segmental_conll2000.py]: The way I'm doing this is just to override `_max_train_sentence_len` when I run evaluate. This more robust than this hack. Particularly because it's really obvious when you're modifying the default.

File: allennlp/data/dataset_readers/segmental_conll2000.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                         tokens: List[Token],
+                         chunk_tags: List[str] = None) -> Instance:
+        """
+        We take `pre-tokenized` input here, because we don't have a tokenizer in this class.
+        """
+        # pylint: disable=arguments-differ
+        sentence = TextField(tokens, self._token_indexers)
+        instance_fields: Dict[str, Field] = {'tokens': sentence}
+
+        if chunk_tags is not None:
```
Comment (brendan-ai2) [inline allennlp/data/dataset_readers/segmental_conll2000.py]: Another place for flattening. Again, just a tip! Something like:

```
if chunk_tags is None:
    return Instance({})

chunk_tags = ...
```

**Comment by brendan-ai2** [comment]
Gave this a quick skim. Looks good! Left a few suggestions, but since this is your fork, feel free to treat them as just that!

File: allennlp/commands/evaluate.py (hunk 1 above)
Comment (swabhs) [reply allennlp/commands/evaluate.py]: Fixed, thanks!

File: allennlp/data/dataset_readers/segmental_conll2000.py (hunk 2 above)
Comment (swabhs) [reply allennlp/data/dataset_readers/segmental_conll2000.py]: Makes sense, fixed!

File: allennlp/data/dataset_readers/segmental_conll2000.py (hunk 3 above)
Comment (swabhs) [reply allennlp/data/dataset_readers/segmental_conll2000.py]: I agree your suggestion is much better than this hack, but  I wrote this for the case where I was using a validation set during training, where that logic won't work?

**Comment by swabhs** [comment]
Thanks a lot @brendan-ai2!

File: allennlp/data/dataset_readers/segmental_conll2000.py (hunk 3 above)
Comment (brendan-ai2) [reply allennlp/data/dataset_readers/segmental_conll2000.py]: Oh, interesting! I actually wasn't using a validation set as the sampled softmax code defaulted to doing a real softmax when not training which caused OOMs. I'm curious, what are you using as your validation set?

Comment (swabhs) [reply allennlp/data/dataset_readers/segmental_conll2000.py]: Sorry this was before I started working on the 1b benchmark, and was using PTB to build LMs.
